### ISSUE TITLE
ZFS ignores all IOCB_* flags

### ISSUE THREAD

**Issue by Uglymotha** [zfsonlinux/zfs#8452]
Distribution Name	|  N/A
Distribution Version	|  N/A 
Linux Kernel	|  4.19 and above
Architecture	|  x64
ZFS Version	|  0.7.12
SPL Version	|  0.7.12

### Describe the problem you're observing
In a bug discussion wirth the developer of SCST target the following came to light, this may or may not have bee resolved in upstream or yet to be release version, please treat at your discretion:

Regarding the combination of asynchronous file I/O and ZFS: I expect 
that making the SCST vdisk_fileio handler submit asynchronous direct I/O 
to the underlying filesystem will yield a performance improvement. SCST 
tells the underlying filesystem about the I/O type by setting the 
IOCB_DIRECT and IOCB_NOWAIT flags in struct kiocb. The results of a 
quick analysis of how ZFS handles these IOCB_* flags is as follows:

$ git grep IOCB_ | wc -l
       0

In other words, ZFS ignores all IOCB_* flags. I think this is a severe 
bug. The drivers I am aware of that set this flag are (kernel v5.0-rc7 / 
SCST trunk):
* File I/O from user space for files opened with O_DIRECT.
* The asynchronous I/O code for files opened with O_DIRECT
   (e.g. fio --ioengine=libaio --direct=1).
* The loop driver.
* The NVMe target driver in direct I/O mode.
* The LIO fileio driver with async mode enabled.
* The SCST vdisk_fileio handler.

### Describe how to reproduce the problem
N/A

**Comment by behlendorf** [comment]
@Uglymotha thank you for filing this issue.  We're looking in to the best way to handle this.
